fix: delete node_modules if frappe supports app_cache

18alantom opened the pull request:
Since part of the build executes in frappe, get-app cache requires at least v15.12.0 of frappe to function optimally i.e. delete node_modules and skip build. This PR checks frappe version mentioned in the app's `pyproject.toml`.


Uses version specifier format mentioned in: https://github.com/frappe/bench/issues/1524

18alantom committed:
fix: del node_modules if frappe supports app_cache

ankush commented:
@18alantom just few steps away from closing linked issue, might as well complete it :grimacing:

18alantom commented:
> @18alantom just few steps away from closing linked issue, might as well complete it 😬

@ankush I'm unsure about what all needs to be done w.r.t the linked issue:
- How should dependency absence be handled, fail get-app or just show a warning?
- Assuming `frappe-dependencies` are runtime, i.e. get-app can still run without the `frappe-dependencies`, doesn't make sense to fail it.
- Any way to indicate strict requirement during get-app, (in case `setup.py` needs the dependency).

If we're looking to fail get-apps if mentioned dependency is not present, we should hold off until more thought is given. Press Deploy Candidate app ordering too would have to be updated to account for it.

ankush commented:
Yes we should fail with non-zero exit code.

If ordering is a hard problem to solve we can optionally add a check command to do it after all apps are installed on bench?

18alantom committed:
fix: check frappe-deps during get-app

18alantom commented:
Added dependency check, fails if folder not present. Also added a version check, doesn't fail but gives a warning if mismatch.

Regarding ordering:
1. Ordering needs to be managed by whomever is installing apps using bench.
2. If (all) apps have frappe-dependencies mentioned, invalid ordering is not possible (cause it will fail).
3. If apps do not have frappe-dependencies mentioned, invalid ordering cannot be prevented (else get-app breaks).
4. By point 2. and 3. ordering check in bench is redundant.
